### PR TITLE
Reboot

### PR DESCRIPTION
**Pull request by erp12**
This pull request is best reviewed 1 commit at time. When making some of the changes we discussed (specifically moving away from records) many files were touched and a lot of code needed to change to fix the failing tests. 

451f762, b8adc11, and aa302a7 are simple changes that mostly deal with project/repository setup.

1d31382 is a nasty commit that really should rebased into multiple smaller ones. I had trouble organizing my changes here because I kept getting confused while trying to debug my initial changes, which prompted me to reorganize more code and in the end a lot of code got touched. It's a pretty nasty diff to review. Here is the outline of what was done:

- Changes Push units to maps instead of records.
- Changes type-set to type-library.
- Adds print trace option to program execution.
- Moves instruction type definitions to a separate namespace.
- Push programs hold `instruction-meta` objects instead of entire instructions. During program execution, instruction-meta objects are used to lookup the correct instruction in the instruction set.

4626a8c is where the instruction testing code was introduced.

### CODE REVIEW

**Comment by Vaguery** [comment]
Remind me _why_ we moved away from records to maps? I lean towards using them since there are no functional differences, and more convenient constructors and checkers, and very little (if any) overhead.

Also, I'm not sufficiently competent to rebase and collapse 1d31382. If you want it done, I will have to ask for hand-holding.

**Review by Vaguery — APPROVED** [review]

File: src/puj/push/instruction.clj [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+(s/def ::stack-list (s/coll-of ::stack-name))
+
+(s/def ::touched-stacks ::stack-list)
+
+(s/def ::input-stacks ::stack-list)
+
+(s/def ::output-stacks ::stack-list)
+
+(s/def ::other-stacks ::stack-list)
```
Comment (Vaguery) [inline src/puj/push/instruction.clj:23]: Have to ask for an example here. Are there instructions with "side-effects", which I assume this is for, or is it "just in case"? I would rather not include it until we have a use case.

File: src/puj/push/instructions/numeric.clj [hunk 2]
```diff
@@ -1,31 +1,28 @@
...
      {:puj.push.instruction/doc  "Pushes the sum of the top 2 ints onto the int stack."})
 
-   :int-sub
    (with-meta
-     (u/simple-instruction [:int :int] :int 0 -)
+     (i/simple-instruction :int-sub [:int :int] :int 0 -)
      {:puj.push.instruction/doc  "Pushes the difference of the top 2 ints onto the int stack."})
 
-   :int-mult
    (with-meta
```
Comment (Vaguery) [inline src/puj/push/instructions/numeric.clj:20]: After about three of these, I would like to suggest we clean up the `with-meta` junk if possible. If we're going to be wrapping all these definitions in _other_ definitions (so they have docstrings), then we should consider a simpler function or macro that produces a documented function, not make every new instruction have this layered structure.

**Review by Vaguery — APPROVED** [review]

**Comment by lspector** [comment]
> Remind me _why_ we moved away from records to maps? I lean towards using them since there are no functional differences, and more convenient constructors and checkers, and very little (if any) overhead.

I can't speak for others but I'm happy to see them go for a couple of reasons, including stuff that stems from them sort-of-being-but-also-not-being maps. I know there are benefits, but I've been burned by the weaknesses several times. Oh and apparently there are some special weirdnesses across the clj/cljs divide (described [here](https://blog.klipse.tech/clojurescript/2016/04/25/records-wacky-maps.html) along with some other issues).

File: src/puj/push/instruction.clj (hunk 1 above)
Comment (erp12) [reply src/puj/push/instruction.clj:23]: This is common in the instructions of PyshGP and Clojush. For example, `string-stack-depth` requires a string stack to exist, but it doesn't consume or push items to it.

File: src/puj/push/instructions/numeric.clj (hunk 2 above)
Comment (erp12) [reply src/puj/push/instructions/numeric.clj:20]: I am not sure what you mean by layering, but I am fine with moving docstrings out of metadata.
We could keep a separate namespace with a map of instruction names to docstrings, but there is something to be said for having the docstring exist in a line that is nearby the instruction definition.

Given that docstrings are logically optional metadata, it seems like a natural implementation to me. We use qualified keywords that won't conflict with metadata from other libraries. Can you elaborate more on why you see this as a concern?

**Comment by erp12** [comment]
> Remind me _why_ we moved away from records to maps? I lean towards using them since there are no functional differences, and more convenient constructors and checkers, and very little (if any) overhead.

I agree with @lspector and can add a few other points.

Personally, I find the constructors for records to be a hindrance rather than a benefit. They cannot be customized and I always end up wrapping them. Even if you don't need to customize the constructor, you might in the future. This means you will break callers when you inevitably wrap your constructor in a function. For example, adding a default value to an argument. This is a minor increase in code, but if we never call the constructors directly, their existance isn't helping us much.

My bigger issue is the protocols. I have come to the opinion that there is very little reason to use them outside of interop. It is annoying to have a set of functions that work on a record but not the equivalent map. You have to guard against your data losing it's record status by repeatedly casting back. I know we can use Records without Protocols, that pattern generally results in your data quickly reverting to a map anyways.

There is also a nasty interaction with records and specs. Consider this data model from this PR:

```clj
(s/def ::instruction-meta
  (s/keys :req [::name ::code-blocks]))

(s/def ::instruction-base
  (s/and ::instruction-meta
         (s/keys :req [::logic-fn])))

(s/def ::simple-instruction
  (s/and ::instruction-base
         (s/keys :req [::input-stacks ::output-stacks])))
```

The following map is a `instruction-meta`.

```clj
{::name :int-add 
 ::code-blocks 0}
```

To turn that into a simple instruction, we add fields.

```clj
{::name :int-add 
 ::code-blocks 0
 ::logic-fn +
 ::input-stacks [:int :int]
 ::output-stacks :int}
```

If these were all records, we would only have 2 ways to do this kind of conversion.

1. Round trip from `InstructionMeta` record to a map, assoc, and then create a `SimpleInstruction` record. No benefits there.

2. We could implement methods of the `InstructionMeta` that takes the missing fields as arguments and directly creates the `SimpleInstruction` record. Its very fast, but its a very narrow API. You can _only_ turn an `InstructionMeta` into a `SimpleInstruction` and you can _only_ add values for the fields of `SimpleInstructions`. I would rather use Clojure awesome generic map-manipulation API unconstrained rather than implement dozens of narrow methods on our records.

The TL;DR version is: spec + maps is open (put anything you want in!), records are closed (don't stray, or you are back to maps).

I know the most common trade-off is performance, because field look-ups are cheaper on a record. If/when performance ends up being an issue, we can profile the impact of maps vs records and make a surgical change to address the issue. Until then, I would prefer simplicity.

**Comment by erp12** [comment]
> Also, I'm not sufficiently competent to rebase and collapse [1d31382](https://github.com/push-language/Puj/commit/1d313826d1659aa75d1be132907a9ce2e70b881b). If you want it done, I will have to ask for hand-holding.

I know it is difficult to review, but unless anyone feels strongly about breaking it up I think we should just leave the commit as-in.

File: src/puj/push/instructions/numeric.clj (hunk 2 above)
Comment (thelmuth) [reply src/puj/push/instructions/numeric.clj:20]: I like this as metadata, unless @Vaguery has a good alternative.

**Comment by thelmuth** [comment]
I'm happy with this! I do not see the GitHub place to say that I've reviewed it though...
